src/com/google/enterprise/adaptor/database/UniqueKey.java

```diff
@@ -160,6 +163,51 @@
     return aclSqlCols;
   }
 
+  /**
+   * Attempt to extract ColumnTypes from the supplied Map
+   * of column names to java.sql.Type
+   *
+   * @param typeMap Map of column names to Integer java.sql.Types
+   */
+  void addColumnTypes(Map<String, Integer> typeMap) {
+    for (Map.Entry<String, Integer> entry : typeMap.entrySet()) {
+      if (types.get(entry.getKey()) != ColumnType.UNKNOWN) {
+        continue;
+      }
+      ColumnType type;
+      switch (entry.getValue()) {
+        case Types.INTEGER:
+        case Types.SMALLINT:
+        case Types.TINYINT:
+          type = ColumnType.INT;
+          break;
+        case Types.BIGINT:
+          type = ColumnType.LONG;
+          break;
+        case Types.CHAR:
+        case Types.VARCHAR:
+        case Types.LONGVARCHAR:
+          type = ColumnType.STRING;
+          break;
+        case Types.DATE:
+          type = ColumnType.DATE;
+          break;
+        case Types.TIME:
+          type = ColumnType.TIME;
+          break;
+        case Types.TIMESTAMP:
+          type = ColumnType.TIMESTAMP;
+          break;
+        default:
+          String errmsg = "Invalid UniqueKey type '" + entry.getValue() + "'"
+              + " for " + entry.getKey() + ". Valid types are: int, string,"
+              + " timestamp, date, time and long.";
+          throw new InvalidConfigurationException(errmsg);
+      }
+      types.put(entry.getKey(), type);
+    }
+  }
+
   String makeUniqueId(ResultSet rs, boolean encode) throws SQLException {
```

Comment:
Future work to add checks for UNKNOWN to the methods accessing types after init?